### DISCUSSION TITLE
Error Level Analysis with IM

### DISCUSSION THREAD

**Discussion by artnose** [ImageMagick/ImageMagick#6781] [Digital Image Processing] [answered]
I'm trying to do Error Level Analysis  on images to detect photo manipulation.
"Error level analysis (ELA) works by intentionally re-saving the image at a known error rate, such as 95%, and then computing the difference between the images. If there is virtually no change, then the cell has reached its local minima for error at that quality level. However, if there is a large amount of change, then the pixels are not at their local minima and are effectively original.”"

An example IM command as given as 
convert input.jpg \( +clone -quality 95 \) -compose difference -composite -auto-level -gamma 1.5 output.png

However, this just gives me a black image.  So I'm wondering if there is a way to get this working, or another command that
might offer some ELA insight?

**Comment by fmw42** [comment]
It may just not be visible to your eyes or the 95% may not do anything for that defect.  To do the error analysis better, use `compare` with some metric such as rmse or ncc, etc.

`compare -metric rmse input.jpg output.jpg difference.png`

See https://imagemagick.org/Usage/compare/

**Comment by artnose** [comment]
Hmm,  I do see something now, (basically input image in red) but not what I was expecting, and it certainly doesn't highlight areas of the image edited.

I was expecting something more like this:
http://www.errorlevelanalysis.com/  (if you roll over the image)

**Reply by fmw42** [reply]
Depends on the image and the defect that you are testing

**Comment by snibgo** [comment]
`quality 95` doesn't change the image. It is a setting that affects how the image will be written. So your command compares an image with itself, and finds there is (of course) no difference.

**Reply by fmw42** [reply]
-quality only changes the image when saving a (JPG) output (or some other compressed image format that uses that argument). So you would need to save a new image to compare with the input if all you want to do is change the JPG quality in order to change the compression.  Other defects would require some other command (such as adding noise), which could be done the way you have above without having to save a new image first.

**Comment by artnose** [comment]
again, you can read the website I posted, I'm just looking for results similar to that (with the sample they give).  But I'm not able to reproduce.

**Reply by fmw42** [reply, marked as answer]
```
convert input.jpg -quality 95 output.jpg
convert input.jpg output.jpg -compose difference -composite -auto-level -gamma 1.5 difference.png
```

You may not need the -gamma.  If the result is not strong enough without it, then add it back and adjust the -gamma value. Or try with it and if too strong, then reduce the value or take it out.

**Reply by artnose** [reply]
Thank you!  that works perfectly.  
Here's an example of an edited photo analysis, clearly showing how this highlights edits!  
To the naked eye, the areas highlighted were not noticable.
https://imgur.com/a/bqjiuh2
